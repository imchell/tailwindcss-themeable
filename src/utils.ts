import { ColorHex } from '.'

export const clamp = (val: number, min: number = 0, max: number = 100) => {
  return Math.min(Math.max(val, min), max)
}

export const hex2rgb = (hex: ColorHex): [number, number, number] => {
  let r, g, b
  if (hex.length === 4) {
    // #RGB
    [r, g, b] = hex.split('').slice(1).map(s => parseInt(s + s, 16) / 255)
  } else {
    // #RRGGBB
    const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex)
    if (result == null) {
      throw new Error(`Please use the #RGB or #RRGGBB format for hex colors, got wrong color hex ${hex}`)
    }
    r = parseInt(result[1], 16)
    g = parseInt(result[2], 16)
    b = parseInt(result[3], 16)
  }
  return [r, g, b]
}

export const hex2HSL = (hex: ColorHex): [number, number, number] => {
  const [r, g, b] = hex2rgb(hex).map(char => char / 255)

  const max = Math.max(r, g, b); const min = Math.min(r, g, b)
  let h = 0; let s = 0; const l = (max + min) / 2
  if (max === min) {
    h = s = 0 // achromatic
  } else {
    const d = max - min
    s = l > 0.5 ? d / (2 - max - min) : d / (max + min)
    switch (max) {
      case r: h = (g - b) / d + (g < b ? 6 : 0); break
      case g: h = (b - r) / d + 2; break
      case b: h = (r - g) / d + 4; break
    }
    h /= 6
  }
  return [
    h * 360,
    s * 100,
    l * 100
  ]
}

export const hsl2Hex = ([h, s, l]: [number, number, number]): ColorHex => {
  s = clamp(s, 0, 100)
  l = clamp(l, 0, 100)

  l /= 100
  const a = s * Math.min(l, 1 - l) / 100
  const f = (n: number) => {
    const k = (n + h / 30) % 12
    const color = l - a * Math.max(Math.min(k - 3, 9 - k, 1), -1)
    return Math.round(255 * color).toString(16).padStart(2, '0') // convert to Hex and prefix "0" if needed
  }
  return `#${f(0)}${f(8)}${f(4)}`
}
