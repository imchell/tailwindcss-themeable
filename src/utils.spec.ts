import { hex2HSL, hsl2Hex } from './utils'

describe('utils', () => {
  it('should convert hex to hsl correctly', () => {
    expect(hex2HSL('#FF0000')).toMatchSnapshot();
    ([
      '#FFFFFF', '#EFFFF3',
      '#C7FDD5', '#9FFCB7',
      '#78FB99', '#50FA7B',
      '#19F852', '#06D43A',
      '#049D2B', '#03671C'
    ] as const).map(hex => expect(hex2HSL(hex)).toMatchSnapshot())
  })

  it('should convert hsl to hex correctly', () => {
    ([
      '#FFFFFF', '#EFFFF3',
      '#C7FDD5', '#9FFCB7',
      '#78FB99', '#50FA7B',
      '#19F852', '#06D43A',
      '#049D2B', '#03671C'
    ] as const).map(hex => expect(hex).toStrictEqual(
      hsl2Hex(hex2HSL(hex)).toUpperCase()
    ))
    expect(hsl2Hex(hex2HSL('#FFF')).toUpperCase()).toStrictEqual('#FFFFFF')
    expect(hsl2Hex(hex2HSL('#ABC')).toUpperCase()).toStrictEqual('#AABBCC')
  })
})
