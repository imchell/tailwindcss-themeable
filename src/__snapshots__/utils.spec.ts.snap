// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils should convert hex to hsl correctly 1`] = `
Array [
  0,
  100,
  50,
]
`;

exports[`utils should convert hex to hsl correctly 2`] = `
Array [
  0,
  0,
  100,
]
`;

exports[`utils should convert hex to hsl correctly 3`] = `
Array [
  134.99999999999994,
  100,
  96.86274509803921,
]
`;

exports[`utils should convert hex to hsl correctly 4`] = `
Array [
  135.55555555555557,
  93.10344827586209,
  88.62745098039215,
]
`;

exports[`utils should convert hex to hsl correctly 5`] = `
Array [
  135.48387096774192,
  93.93939393939395,
  80.58823529411765,
]
`;

exports[`utils should convert hex to hsl correctly 6`] = `
Array [
  135.1145038167939,
  94.24460431654677,
  72.74509803921568,
]
`;

exports[`utils should convert hex to hsl correctly 7`] = `
Array [
  135.1764705882353,
  94.44444444444446,
  64.70588235294117,
]
`;

exports[`utils should convert hex to hsl correctly 8`] = `
Array [
  135.33632286995515,
  94.09282700421942,
  53.529411764705884,
]
`;

exports[`utils should convert hex to hsl correctly 9`] = `
Array [
  135.14563106796115,
  94.49541284403668,
  42.74509803921569,
]
`;

exports[`utils should convert hex to hsl correctly 10`] = `
Array [
  135.2941176470588,
  95.03105590062113,
  31.56862745098039,
]
`;

exports[`utils should convert hex to hsl correctly 11`] = `
Array [
  135,
  94.33962264150942,
  20.784313725490197,
]
`;
